### ISSUE TITLE
Titanium code editor with built-in TiShadow/CLI/Alloy support/Code completion

### ISSUE THREAD

**Issue by EtnaTraining** [TiForward/discuss#9]
I think there is the need to have a integrated environment to build Titanium app.
My idea would be to take one of the popular text editor and add support for Titanium development. 
A set of provided feature would be:
- code completion for Titanium JS API, Alloy XML and CSS tags
- launch the different simulators 
- integrate with TiShadow in a way to support live code reload
- provide shortcuts to generate Alloy controllers, models, views, styles

I am currently using Sublime Text 2/3, but code completion is far from perfect. No completion plugins for Alloy. There is a ti-build and tishadow plugin.

Anyway, Sublime Text is not free and open source, and Python is needed to extend it or develop packages.
Other options, more interesting IMHO, are ATOM and Brackets editor.
I don't know much about Brackets, but as far as I have seen, Atom would be the best choice:
- open source
- based on Chromium
- can be extended with JavaScript, CSS, HTML
- give full access to Node.js APIs (file system, networkings, etc)

If you are following Facebook recently, React Native team is building their own IDE, Nuclide, based on Atom:
http://nuclide.io

Anyone have any experience developing Atom packages?
Anyone interested in this activities?

Other options and/or considerations?

**Comment by mattapperson** [comment]
Atom packages are just nodejs based plugins working with their API. 
Building out IDE features is a big task. It would be great to have a nice Ti plugin... but doing so in a way that did not feel like a side project is a job and a half :) 

Also, if adobe Theseus could be made to work with Ti, that would be great. As would AmokJS.

**Comment by mattapperson** [comment]
I will also add that Ti.Forward is NOT a fork of Ti, it is an effort to give back to titanium. And Appc already has an IDE :) 

That said, we are happy to facilitate all conversations regarding titanium :)

**Comment by EtnaTraining** [comment]
Yes, I am not sure how many Titans are happy with Titanium/Appcelerator Studio! Many of us use Sublime Text, instead of Ti.Studio! Have a look to the latest video on Appc University, made by @fokkezb: he shows his code in sublime :) And Jeff himself showed demo with Sublime too several times...

So if the community would give back to Titanium a better editor, everybody would be happy :)

**Comment by EtnaTraining** [comment]
I have never built an Atom packages, but maybe developing several packages, each of them covering a single feature, could be a smaller task, and in the future they could be merged in a single and comprehensive one....

**Comment by mattapperson** [comment]
@EtnaTraining Oh I 100% agree. I am just saying that I doubt Appc would accept a PR for a second IDE (via a massive plugin). Plugins for other editors for things like code completion? This I see as something they might accept. Thats largely up to them, but we can try :)

**Comment by TheBrousse** [comment]
I see it more as "an open-source alternative" to studio rather than a competing product. Remember that Appcelerator Studio, while free, is not open-source, therefore there is no 100% OSS alternative out there. So I think an IDE such as this one could exist as a standalone project since it has no goal of being merged into the main truck.
Although I also agree with Matt when he mentions that developing a full fledge IDE is a huge endeavour. And that a more step-by-step approach should be taken in this particular case.

**Comment by iotashan** [comment]
> Remember that Appcelerator Studio, while free, is not open-source, therefore there is no 100% OSS alternative out there.

Appcelerator Studio is free to try. Once you're ready to deploy an app to the stores, you must sign up for a paid plan. Having a free alternative (better than the basic code coloring that's out there now) is a good idea.

**Comment by Sophrinix** [comment]
I use Appcelerator Studio, Textmate, and VIM. If [Atom could learn to handle big files.. which to their credit they seem to care about](https://github.com/atom/atom/issues/307), I would no longer need Textmate ( even though it is also open source).

**Comment by FokkeZB** [comment]
@iotashan It's not just Studio you pay for, it's also using the certified SDKs to build and the Analytics/Arrow to backend.

But since Appc Studio uses the CLI underneath there would be no reason you can do the same for Atom or Sublime or any other editor.

**Comment by TheBrousse** [comment]
True, therefore a standalone, Titanium only IDE would make sense.
Date: Mon, 13 Apr 2015 10:37:38 -0700

**Comment by EtnaTraining** [comment]
thanks for all the feedback!
Just to clarify I have no intention to propose the development of an IDE!!! Neither a full code editor!
The idea was to develop or integrate existing modules to some existing open source product, to have a free and open source, hackable alternative to Appc Studio (that is the full IDE with debugging support).

**Comment by EtnaTraining** [comment]
from my knowledge and your answers, the list of possibile candidates are:
- atom
- brackets
- textmate

(I have excluded vim/vi because I don't think it's a good choice for a newbie)

Am I missing any open source editors that are customisable?

**Comment by iotashan** [comment]
Aye, yes, you get more than just Appcelerator Studio with a paid account, but conversely, you can't use Appcelerator Studio for production without a paid account.

**Comment by FokkeZB** [comment]
@iotashan true, just wanted to point that out ;)

**Comment by jhaynie** [comment]
We are investing in Studio because we have a lot of users that use the product and because a lot of customers want a full blown IDE.  But we also want to support lighter weight code editors like Sublime, Atom, etc.  So in my opinion, these are not in conflict at all.

**Comment by mattapperson** [comment]
@jhaynie :+1: Great to hear!

**Comment by yomybaby** [comment]
Recently I use Atom. So I realy want to atom packege for titanium including autocomplete and more. Let's making some scratch!

**Comment by dbankier** [comment]
For those wanting code completion for other editors have a look at TernJS and this https://github.com/x-cray/tern-titanium.
